container/shim/src/fetchers/lassie.js

```diff
@@ -1,3 +1,143 @@
+import http from "node:http";
+import https from "node:https";
+
+import { CarBlockIterator } from "@ipld/car";
+import LRU from "lru-cache";
+import write from "stream-write";
+
+import { LASSIE_ORIGIN } from "../config.js";
+import { streamCAR, validateCarBlock } from "../utils/car.js";
+import { proxyResponseHeaders, toUtf8 } from "../utils/http.js";
+import { debug as Debug } from "../utils/logging.js";
+
+const debug = Debug.extend("lassie");
+const LASSIE_TIMEOUT = 120_000;
+
+const agentOpts = {
+  keepAlive: true,
+};
+const httpsAgent = new https.Agent(agentOpts);
+const httpAgent = new http.Agent(agentOpts);
+
+const blockCache = new LRU({
+  maxSize: 1024 ** 3, // 1 GB
+  sizeCalculation: (block) => Buffer.byteLength(block),
+  allowStale: true,
+});
+
 export function respondFromLassie(req, res, { cid, format }) {
-  // todo
+  debug(`Fetch ${req.path} from Lassie`);
+
+  const isRawFormat = format === "raw";
+  const isInBlockCache = isRawFormat && !req.params.path && blockCache.has(cid);
+  if (isInBlockCache) {
+    const block = blockCache.get(cid);
```

Comment:
For these cache operations, get and put, do a `cid.toV1().toString()`. Since you only care about the bytes all that really matters is the multihash (you could even extract that and use that as the key) and you want to dedupe content where possible. It's not super-rare that you'll see the same content addressed by a v1 and a v0 CID but they're the same content.
Doing it by multihash is even better because you're just returning bytes out of this and it doesn't matter what the rest of the CID says in that case. There is a bunch of content stored in Filecoin now as "raw" that is really something else; most of our blockstores address by multihash now so it shouldn't matter from a source-of-block-data perspective.

Reply:
If `cid.toV1().toString()` works fine then I'd prefer that for DX, since the cache requires string keys. I'd rather look at CID keys than stringified bytes.

Reply:
Yep, that's fair enough and probably what I'd do for a first pass at this.

Just so you're aware though, you could also be doing a CID-style multibase encoding of the multihash and it would probably be as cheap, perhaps cheaper than doing the CID stringification and you can do it in a more compact string form too to save bytes in the LRU's key storage. You'd also then be properly disconnecting codec and CID version from the content being requested - if upstream wants bytes-for-hash then you just serve them back, regardless of the codec because that's not your problem.

<img width="965" alt="Screenshot 2023-02-08 at 10 38 58 am" src="https://user-images.githubusercontent.com/495647/217391733-37e87cf1-b9ef-4c83-b9bc-7f44a4e55343.png">

(`baseEncode` leaves off the multibase prefix since it's not helpful if they're all the same base, and adds a string concat operation; `encode` would put the `m` on there for base64 multibase).

Reply:
Ok you've convinced me now that I've extracted a code example from you 😛